components/TextFields/src/Private/MDCTextInputCommonFundament.m

```diff
@@ -0,0 +1,944 @@
+  return [UIColor colorWithWhite:0 alpha:MDCTextInputHintTextOpacity];
+}
+
+static inline UIColor *MDCTextInputTextColor() {
+  return [UIColor colorWithWhite:0 alpha:[MDCTypography body1FontOpacity]];
+}
+
+static inline UIColor *MDCTextInputUnderlineColor() {
+  return [UIColor lightGrayColor];
+}
+
+static inline CGFloat MDCRound(CGFloat value) {
+#if CGFLOAT_IS_DOUBLE
+  return rint(value);
+#else
+  return rintf(value);
+#endif
+}
+
+@interface MDCTextInputCommonFundament () {
+  BOOL _mdc_adjustsFontForContentSizeCategory;
+}
+
+@property(nonatomic, strong) UIImage *clearButtonImage;
+@property(nonatomic, strong) NSLayoutConstraint *clearButtonWidth;
+@property(nonatomic, assign) BOOL isRegisteredForKVO;
+@property(nonatomic, strong) NSLayoutConstraint *placeholderHeight;
+@property(nonatomic, strong) NSLayoutConstraint *placeholderLeading;
+@property(nonatomic, strong) NSLayoutConstraint *placeholderLeadingLeftViewTrailing;
+@property(nonatomic, strong) NSLayoutConstraint *placeholderTop;
+@property(nonatomic, strong) NSLayoutConstraint *placeholderTrailing;
+@property(nonatomic, strong) NSLayoutConstraint *placeholderTrailingRightViewLeading;
+@property(nonatomic, strong) UIView *relativeSuperview;
+@property(nonatomic, weak) UIView<MDCTextInput> *textInput;
+@property(nonatomic, strong) NSLayoutConstraint *underlineY;
+
+@end
+
+@implementation MDCTextInputCommonFundament
+
+// We never use the text property. Instead always read from the text field.
+
+@synthesize attributedText = _do_no_use_attributedText;
+@synthesize clearButton = _clearButton;
+@synthesize clearButtonColor = _clearButtonColor;
+@synthesize clearButtonMode = _clearButtonMode;
+@synthesize hidesPlaceholderOnInput = _hidesPlaceholderOnInput;
+@synthesize leadingUnderlineLabel = _leadingUnderlineLabel;
+@synthesize placeholderLabel = _placeholderLabel;
+@synthesize positioningDelegate = _positioningDelegate;
+@synthesize textColor = _textColor;
+@synthesize trailingUnderlineLabel = _trailingUnderlineLabel;
+@synthesize underline = _underline;
+
+- (instancetype)init {
+  [self doesNotRecognizeSelector:_cmd];
+  return nil;
+}
+
+- (nonnull instancetype)initWithTextInput:(UIView<MDCTextInput> *_Nonnull)textInput {
+  self = [super init];
+  if (self) {
+    _textInput = textInput;
+
+    [self commonMDCTextInputCommonFundamentInit];
+    self.textInput.font = [UIFont mdc_preferredFontForMaterialTextStyle:MDCFontTextStyleBody1];
+
+    // Initialize elements of UI
+    [self setupRelativeSuperview];
+
+    [self setupPlaceholderLabel];
+    [self setupUnderlineView];
+    [self setupClearButton];
+    [self setupUnderlineLabels];
+
+    [self updateColors];
+    [self mdc_setAdjustsFontForContentSizeCategory:NO];
+
+    [self subscribeForKVO];
+  }
+  return self;
+}
+
+- (instancetype)initWithCoder:(NSCoder *)aDecoder {
+  UIView<MDCTextInput> *input = [aDecoder decodeObjectForKey:MDCTextInputCoordinatorInputKey];
```

Comment:
Is there a reason this is done before self = [super init] ?
Consider moving into the if clause to be consistent with the other decoder fields.